trie/smt.py

```diff
@@ -0,0 +1,266 @@
+        return self._key
+
+    @property
+    def value(self):
+        return self._value
+
+    @property
+    def branch(self):
+        return self._branch
+
+    @property
+    def root_hash(self):
+        return calc_root(self.key, self.value, self.branch)
+
+    def merge(self, key, value, node_updates):
+        """
+        Algorithm takes two separate entries in an SMT and merges them.
+        This does not require access to the original SMT to perform
+        if branch is merged for a stream of all updates that have happened.
+        """
+        validate_is_bytes(key)
+
+        # Validate that the two keys are the same size
+        validate_length(key, self._keysize)
+
+        # Path diff is the logical XOR of the updated key and this account
+        path_diff = (to_int(self.key) ^ to_int(key))
+
+        # Full match to key (no diff), update the key's value instead
+        if path_diff == 0:
+            self._value = value
+            # No need to update branch
+        else:
+            # Find the first mismatched bit between keypaths. This is
+            # where the branch point occurs, and we should update the
+            # sibling node in the source branch at the branch point.
+            # NOTE: Keys are in MSB->LSB (root->leaf) order.
+            #       Node lists are in root->leaf order.
+            #       Be sure to convert between them effectively.
+            for bit in reversed(range(self._branchsize)):
+                if path_diff & (1 << bit) > 0:
+                    branch_point = (self._branchsize-1)-bit
+                    break
+
+            # NOTE: node_updates only has to be as long as necessary
+            #       to obtain the update. This allows an optimization
+            #       of pruning updates to the maximum possible depth
+            #       that would be required to merge, which may be many
+            #       orders of magnitude smaller than the tree depth.
+            if len(node_updates) <= branch_point:
+                raise ValidationError("Updated node list is not deep enough")
+
+            # Update sibling in the branch where our key differs from the update
+            self.branch[branch_point] = node_updates[branch_point]
+            # No need to update leaf
+
+
+class SparseMerkleTree:
+
+    def __init__(self, keysize=32, default=BLANK_NODE):
+        # Ensure we can support the given depth
+        if not 1 <= keysize <= 32:
+            raise ValidationError("Keysize must be number of bytes in range [0, 32]")
+
+        self.keysize = keysize  # keysize in bytes
+        self.depth = keysize * 8  # depth is number of bits in key
+
+        self._default = default
+
+        # Initialize an empty tree with one branch
+        self.db = {}
+        node = self._default  # Default leaf node
+        for i in range(self.depth):
+            node_hash = keccak(node)
+            self.db[node_hash] = node
+            node = node_hash + node_hash
```

Comment:
So... should I do anything to make this better?

Reply:
> So... should I do anything to make this better?

I don't think so.  Most of the time when I see mutation I like to experiment with what it takes to remove it.  Often it isn't worth it (like this case) but I still like dumping the answer in case it's educational to anyone for how to do certain things using an functional approach.